File: delivery/http/handler_auth.go
package http

import (
	"github.com/gofiber/fiber/v2"
	"github.com/teledrive/teledrive/usecase"
)

// AuthHandler handles authentication-related HTTP requests.
type AuthHandler struct {
	authUsecase *usecase.AuthUsecase
}

// NewAuthHandler creates a new AuthHandler.
func NewAuthHandler(authUsecase *usecase.AuthUsecase) *AuthHandler {
	return &AuthHandler{authUsecase: authUsecase}
}

// sendCodeRequest is the JSON body for the send-code endpoint.
type sendCodeRequest struct {
	Phone string `json:"phone"`
}

// HandleSendCode handles POST /api/auth/send-code.
// It initiates the Telegram authentication flow by sending a verification code.
func (h *AuthHandler) HandleSendCode(c *fiber.Ctx) error {
	var req sendCodeRequest
	if err := c.BodyParser(&req); err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error": "invalid request body",
		})
	}

	if req.Phone == "" {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error": "phone number is required",
		})
	}

	result, err := h.authUsecase.SendCode(req.Phone)
	if err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error": err.Error(),
		})
	}

	return c.Status(fiber.StatusOK).JSON(fiber.Map{
		"transaction_id": result.TransactionID,
	})
}

// signInRequest is the JSON body for the sign-in endpoint.
type signInRequest struct {
	TransactionID string `json:"transaction_id"`
	Code          string `json:"code"`
	Password      string `json:"password,omitempty"`
}

// HandleSignIn handles POST /api/auth/sign-in.
// It completes the authentication flow using the verification code.
func (h *AuthHandler) HandleSignIn(c *fiber.Ctx) error {
	var req signInRequest
	if err := c.BodyParser(&req); err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error": "invalid request body",
		})
	}

	if req.TransactionID == "" {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error": "transaction_id is required",
		})
	}
	if req.Code == "" {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error": "code is required",
		})
	}

	result, err := h.authUsecase.SignIn(req.TransactionID, req.Code, req.Password)
	if err != nil {
		return c.Status(fiber.StatusUnauthorized).JSON(fiber.Map{
			"error": err.Error(),
		})
	}

	return c.Status(fiber.StatusOK).JSON(fiber.Map{
		"token": result.Token,
		"user":  result.User,
	})
}

// updateProfileRequest is the JSON body for the update-profile endpoint.
type updateProfileRequest struct {
	FirstName string `json:"first_name"`
	LastName  string `json:"last_name"`
	Phone     string `json:"phone"`
	PhotoURL  string `json:"photo_url"`
}

// HandleUpdateProfile handles PUT /api/auth/profile.
func (h *AuthHandler) HandleUpdateProfile(c *fiber.Ctx) error {
	userID, err := getUserID(c)
	if err != nil {
		return c.Status(fiber.StatusUnauthorized).JSON(fiber.Map{
			"error": "unauthorized",
		})
	}

	var req updateProfileRequest
	if err := c.BodyParser(&req); err != nil {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error": "invalid request body",
		})
	}

	if req.FirstName == "" {
		return c.Status(fiber.StatusBadRequest).JSON(fiber.Map{
			"error": "first_name is required",
		})
	}

	user, err := h.authUsecase.UpdateProfile(userID, req.FirstName, req.LastName, req.Phone, req.PhotoURL)
	if err != nil {
		return c.Status(fiber.StatusInternalServerError).JSON(fiber.Map{
			"error": err.Error(),
		})
	}

	return c.Status(fiber.StatusOK).JSON(user)
}

File: src/delivery/http/handler-auth.ts
import type { Request, Response, NextFunction } from 'express';
import type { AuthUsecase } from '../../usecase/auth.usecase.js';
import { getAuthUserId } from './middleware.js';

export class AuthHandler {
  private authUsecase: AuthUsecase;

  constructor(authUsecase: AuthUsecase) {
    this.authUsecase = authUsecase;
  }

  public handleSendCode = async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    try {
      const { phone } = req.body || {};
      if (!phone || typeof phone !== 'string') {
        res.status(400).json({ error: 'phone number is required' });
        return;
      }

      const result = await this.authUsecase.sendCode(phone);
      res.status(200).json(result);
    } catch (err: any) {
      res.status(400).json({ error: err.message || 'Failed to send code' });
    }
  };

  public handleSignIn = async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    try {
      const { transaction_id, code, password } = req.body || {};
      if (!transaction_id) {
        res.status(400).json({ error: 'transaction_id is required' });
        return;
      }
      if (!code) {
        res.status(400).json({ error: 'code is required' });
        return;
      }

      const result = await this.authUsecase.signIn(transaction_id, code, password);
      res.status(200).json(result);
    } catch (err: any) {
      res.status(401).json({ error: err.message || 'Sign in failed' });
    }
  };

  public handleUpdateProfile = async (req: Request, res: Response, next: NextFunction): Promise<void> => {
    try {
      const userId = getAuthUserId(req);
      const { first_name, last_name, phone, photo_url } = req.body || {};

      if (!first_name) {
        res.status(400).json({ error: 'first_name is required' });
        return;
      }

      const updatedUser = await this.authUsecase.updateProfile(
        userId,
        first_name,
        last_name || '',
        phone || '',
        photo_url || ''
      );

      res.status(200).json(updatedUser);
    } catch (err: any) {
      res.status(500).json({ error: err.message || 'Failed to update profile' });
    }
  };
}
